test/Microsoft.DotNet.PackageInstall.Tests/TestToolBuilder.cs

```diff
@@ -0,0 +1,173 @@
+            testProject.AdditionalProperties["ImplicitUsings"] = "enable";
+            testProject.AdditionalProperties["Version"] = toolSettings.ToolPackageVersion;
+
+            if (toolSettings.NativeAOT)
+            {
+                testProject.AdditionalProperties["PublishAot"] = "true";
+                testProject.AddItem("ToolPackageRuntimeIdentifier", "Include", RuntimeInformation.RuntimeIdentifier);
+            }
+
+            testProject.SourceFiles.Add("Program.cs", "Console.WriteLine(\"Hello Tool!\");");
+
+            var testAssetManager = new TestAssetsManager(log);
+            var testAsset = testAssetManager.CreateTestProject(testProject, identifier: toolSettings.GetIdentifier());
+
+            var testAssetProjectDirectory = Path.Combine(testAsset.Path, testProject.Name!);
+
+            // Avoid rebuilding the package unless the project has changed.  If there is a difference in contents between the files from the TestProject and the target directory,
+            // then we delete and recopy everything over.
+            if (!AreDirectoriesEqual(testAssetProjectDirectory, targetDirectory))
+            {
+                if (Directory.Exists(targetDirectory))
+                {
+                    Directory.Delete(targetDirectory, true);
+                }
+
+                Directory.CreateDirectory(Path.GetDirectoryName(targetDirectory)!);
+
+                Directory.Move(testAssetProjectDirectory, targetDirectory);
+            }
+
+            // If the .nupkg hasn't been created yet, then build it.  This may be because this is the first time we need the package, or because we've updated the tests and
+            // the contents of the files have changed
+            string packageOutputPath = Path.Combine(targetDirectory, "bin", "Release");
+            if (!Directory.Exists(packageOutputPath) || Directory.GetFiles(packageOutputPath, "*.nupkg").Length == 0)
+            {
+                new DotnetPackCommand(log)
+                    .WithWorkingDirectory(targetDirectory)
+                    .Execute()
+                    .Should().Pass();
+
+                if (toolSettings.NativeAOT)
+                {
+                    //  For Native AOT tools, we need to repack the tool to include the runtime-specific files that were generated during publish
+                    new DotnetPackCommand(log, "-r", RuntimeInformation.RuntimeIdentifier)
+                        .WithWorkingDirectory(targetDirectory)
+                        .Execute()
+                        .Should().Pass();
+                }
+
+                //  If we have built a new package, delete any old versions of it from the global packages folder
+                RemovePackageFromGlobalPackages(log, toolSettings.ToolPackageId, toolSettings.ToolPackageVersion);
+            }
+
+            return packageOutputPath;
+        }
+
+        public void RemovePackageFromGlobalPackages(ITestOutputHelper log, string packageId, string version)
```

Comment:
In case you're not already aware, the global packages folder can be changed via [nuget.config, with the `globalPackagesFolder` config](https://learn.microsoft.com/en-us/nuget/reference/nuget-config-file#config-section). NuGet does this in the NuGet.Client repo for restore tests. But considering this PR is about tools, I'm not sure if `dotnet tools` loads nuget.config files, and if so, if it takes into account the CWD, or if it looks at the user nuget.config only.

Reply:
I believe we redirect the global packages folder to a local folder, but we may not clean it each time for perf reasons.  So this method uses the NuGet locals command to resolve the correct global packages folder in the current context (and tools do take into account NuGet.config configuration from the current working directory).